backend/canisters/community/impl/src/activity_notifications.rs

```diff
@@ -23,6 +22,96 @@ pub(crate) fn handle_activity_notification(state: &mut RuntimeState) {
         );
     }
 
+    fn extract_activity(now: TimestampMillis, data: &Data) -> PublicCommunityActivity {
+        let one_hour_ago = now - HOUR_IN_MS;
+        let one_day_ago = now - DAY_IN_MS;
+
+        let mut activity = PublicCommunityActivity {
+            timestamp: now,
+            member_count: data.members.len(),
+            channel_count: data.channels.public_channel_count(),
+            ..Default::default()
+        };
+
+        let mut inc_member_count = |count, within_last_hour| {
+            activity.last_day.member_count_change += count;
+            if within_last_hour {
+                activity.last_hour.member_count_change += count;
+            }
+        };
+
+        for event in data.events.iter(None, false).take_while(|e| e.timestamp >= one_day_ago) {
+            let within_last_hour = event.timestamp >= one_hour_ago;
+
+            match &event.event {
+                CommunityEvent::Created(_) => {
+                    inc_member_count(1, within_last_hour);
+                }
+                CommunityEvent::MemberJoined(_) => {
+                    inc_member_count(1, within_last_hour);
+                }
+                CommunityEvent::MembersRemoved(users) => {
+                    let count = users.user_ids.len() as i32;
+                    inc_member_count(-count, within_last_hour);
+                }
+                CommunityEvent::UsersBlocked(users) => {
+                    let count = users.user_ids.len() as i32;
+                    inc_member_count(-count, within_last_hour);
+                }
+                CommunityEvent::MemberLeft(_) => {
+                    inc_member_count(-1, within_last_hour);
+                }
+                _ => {}
+            }
+        }
+
+        let mut message_unique_users = HashSet::new();
+        let mut reaction_unique_users = HashSet::new();
+
+        for event in data
+            .channels
+            .iter()
+            .filter(|channel| channel.chat.is_public)
+            .filter_map(|channel| {
+                channel.chat.events.events_list(EventIndex::default(), None, now).map(|list| {
```

Comment:
I think you can use `main_events_reader` here instead.

Reply:
Having borrowing issues trying to use a reader

Reply:
Let's discuss tomorrow